### PR TITLE
WIP: Adds Adyen::HPP::Signature with support for HMAC-256 method.

### PR DESCRIPTION
**Pull request by prognostikos**
I've added the logic for signing in a new class, and I plan to remove the old signing code from the Adyen::Form module after I implement it in the Signature class as well. WDYT?

### CODE REVIEW

File: test/hpp/signature_test.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+require 'test_helper'
+
+class SignatureTest < Minitest::Test
+  def setup
+    @shared_secret = "4468D9782DEF54FCD706C9100C71EC43932B1EBC2ACF6BA0560C05AAA7550C48"
```
Comment (wvanbergen) [inline test/hpp/signature_test.rb]: I hope this is for a testing account :)

File: lib/adyen/hpp/signature.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+      def secure_compare(a, b)
+        return false unless a.bytesize == b.bytesize
+
+        l = a.unpack "C#{a.bytesize}"
+
+        res = 0
+        b.each_byte { |byte| res |= byte ^ l.shift }
+        res == 0
+      end
```
Comment (wvanbergen) [inline lib/adyen/hpp/signature.rb]: Why are you using this complicated comparison algorithm? Could `a.force_encoding('binary') == b.force_encoding('binary')` work if you are worried about encodings?

**Comment by wvanbergen** [comment]
This looks good :+1:

File: lib/adyen/hpp/signature.rb (hunk 2 above)
Comment (prognostikos) [inline lib/adyen/hpp/signature.rb]: It's a constant time compare to resist timing attacks.

I stole it from https://github.com/rails/rails/blob/master/activesupport/lib/active_support/security_utils.rb

File: lib/adyen/hpp/signature.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      def sign(params, shared_secret)
+        sig = OpenSSL::HMAC.digest(@digest, Array(shared_secret).pack("H*"), string_to_sign(params))
+        params.merge('merchantSig' => Base64.encode64(sig).strip)
+      end
+
+      # Verify the parameters with the given shared secret
+      # @param [Hash] params The set of parameters to verify. Must include a `merchantSig`
+      #   param that will be compared to the secret calculated here.
+      # @param [String] shared_secret The shared secret for signing/verification
+      # @return [Boolean] true if the `merchantSig` in the params matches our calcuated signature
```
Comment (wvanbergen) [inline lib/adyen/hpp/signature.rb]: Typo: `calcuated`

File: lib/adyen/hpp/signature.rb (hunk 2 above)
Comment (wvanbergen) [inline lib/adyen/hpp/signature.rb]: Ah. I don't think there is a security issue here given that this code runs on the client-side, but it won't hurt either :+1:

File: test/hpp/signature_test.rb (hunk 1 above)
Comment (prognostikos) [inline test/hpp/signature_test.rb]: It's from the doc examples ;)

**Comment by prognostikos** [comment]
I've simplified this now that we don't need configuration so it's now a module with two static methods. It can be merged as-is though nothing else is using it, then I'll rebase the hpp branch and continue working there.

File: lib/adyen/hpp/signature.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      def sorted_keys(hash)
+        hash.sort.map{ |el| el[0] }
+      end
+
+      def sorted_values(hash)
+        hash.sort.map{ |el| el[1] }
+      end
+
+      def escape_value(value)
+        value.gsub(':', '\\:').gsub('\\', '\\\\')
```
Comment (wvanbergen) [inline lib/adyen/hpp/signature.rb]: Nitpick: this can probably be done with a single `gsub` call.

**Comment by wvanbergen** [comment]
I have added you as a collaborator. Feel free top merge yourself :)

**Pull request merged**